Feat: auto-complete support for both block properties and their values

tiensonqin opened the pull request:
This PR addressed several parts:
1. Simplify states for the editor modal, including the current `:editor/action` and its corresponding data (e.g. slash position, input options, etc.)
2. Adds auto-complete support for block properties
3. Adds auto-complete support for block property values
4. Each property will be treated as a "block" too, which enables property blocks in the linked references

Demo:
https://www.loom.com/share/27804e1bcd7b4e4bbf71ec14956c42fe

logseq-cldwalker commented on src/main/frontend/util/text.cljs:
```diff
@@ -86,19 +86,21 @@
   "`pos` must be wrapped by `before` and `and` in string `value`, e.g. ((a|b))"
   [value pos before end]
-  (let [before-matches (->> (get-string-all-indexes value before)
+  (let [before-matches (->> (get-string-all-indexes value before true)
```
Would be nice to have option maps instead of booleans as args as they are easier to read and understand e.g. `{:before? true}`. Also allows you to set a default value if you'd like

logseq-cldwalker commented on src/main/frontend/commands.cljs:
```diff
@@ -333,19 +334,22 @@
+                                   (or
+                                    (and s
+                                         (string/ends-with? s "(")
+                                         (or (string/starts-with? last-pattern "((")
```
Would it be useful to have vars for block-ref, page-ref and property in their respective graph-parser namespaces? I could make a task for this and do it soon

tiensonqin replied:
Yes, it makes sense to move them to the graph-parser. Thanks!

logseq-cldwalker replied:
Sounds good. Created https://quire.io/w/Logseq-_team_dedicated_to_sport_of_knowledge/1123/Move_constants_to_graph-parser to do this later

logseq-cldwalker reviewed (changes requested):
@tiensonqin I'm excited for these changes as this has been near the top of my wishlist and will make ontologies a lot easier to use! Left minor feedback but I think the zotero state is the only one that really needs addressing. Two bugs I also encountered while using this with the hiring graph:
- When I autocomplete properties I see "`type" listed as a property. Maybe this is an mldoc bug?
- After entering the `duration` property, the editor unexpectedly errors. The specific error I get is `Error: Cannot compare 55 to 1h10m`, since one property value is an integer and the other is a string. Should the sort in `get-property-values` be adjusted to handle this?

tiensonqin commented on src/main/frontend/state.cljs:
```diff
@@ -589,63 +586,83 @@
+  []
+  (:editor/action @state))
+
 (defn set-editor-show-page-search!
```
Updated

tiensonqin commented on src/main/frontend/state.cljs:
```diff
@@ -589,63 +586,83 @@
 
 (defn set-editor-show-zotero!
   [value]
   (set-state! :editor/show-zotero value))
```
Thanks! Updated.

logseq-cldwalker commented:
@tiensonqin Thanks for addressing the feedback. What do you think about the two bugs I found in the hiring graph?

tiensonqin commented:
> When I autocomplete properties I see "`type" listed as a property. Maybe this is an mldoc bug?

Mldoc has no limits on the property key at the moment, I added ` to the invalid chars when validating a property key, also truncated some invalid chars like `"^(){}`, there could be better ways for setting the rule. Let me know if you have any recommendations.

> After entering the duration property, the editor unexpectedly errors. The specific error I get is Error: Cannot compare 55 to 1h10m, since one property value is an integer and the other is a string. Should the sort in get-property-values be adjusted to handle this?

Since we don't have schema support for properties, and the types for each property's values could be different, I changed that all the values returned by `get-property-values` will be String. In the future, we might need to allow users/plugins to customize the sorting.

logseq-cldwalker reviewed (approved):
Confirmed the fixes for the bugs I encountered. Excited for this! 🚢 ❤️ 🚢

Matt321456 commented:
Please allow to summon auto-complete for property values without having to go through the whole process of typing :: as a trigger, then picka  property, etc.

Will be epecially useful in the case of:
1. templates with property fields picked out beforehand
2. multiple property values separated by commas (would be nice to have auto-complete for each one of these).
3. simply changing the value into something else

logseq-cldwalker commented:
@Matt321456 Feel free to make a feature request in https://discuss.logseq.com/c/feature-requests/7. Use case 2 shouldn't be a problem anymore as multiple property values should be specified using `[[page]]` and `#tag`, both which have autocompletion. Read more at https://docs.logseq.com/#/page/6356e902-3b7b-4cb2-8c3e-6a904c813c40

Matt321456 commented:
@logseq-cldwalker Will do! Thank you. I'm not sure I understand what you mean by specifying multiple property values. I've read the documentation on properties multiple times. It seems obsolete.

logseq-cldwalker commented:
That page was updated yesterday for changes that landed with 0.8.8. The default used to be:
```
my-property:: value1, value 2
```

The default now is:
```
my-property:: [[value1]] [[value2]]
```
I pointed out that the latter already has autocompletion, hence your 2nd use case is already solved

Matt321456 commented:
@logseq-cldwalker I know about autocompletion [[]] but it's more about page recognition (i.e. linking to a pre-existing page if recognized as such). After entering the property name, you autocomplete a property value that is specific to that property and then I hoped Logseq would do the job of auto-linking if applicable. I'm posted a feature request! Thank you for your help. Please let me know if I'm missing something.

logseq-cldwalker commented:
@Matt321456 The latest nightly and the next release autocompletes the link if you've done that elsewhere

Matt321456 commented:
@logseq-cldwalker I meant autocompletion relative to a set of property values. I wouldn't want "author:: [[W]]" to suggest "Whale" just because I have "Whale" as a page somewhere. I want it to only suggest previous values I've assigned to property "author::". I hope that clarifies it. I know it already does that in the context of a single pair "property -> value" but as soon as I start entering multiple property values, it stops autocompleting from that same set. Thank you.
